module/zfs/metaslab.c

```diff
@@ -4697,7 +4750,7 @@ static void
 metaslab_group_alloc_increment(spa_t *spa, uint64_t vdev, int allocator,
     int flags, uint64_t psize, const void *tag)
 {
-	if (!(flags & METASLAB_ASYNC_ALLOC))
+	if (!(flags & METASLAB_ASYNC_ALLOC) || tag == NULL)
```

Comment:
Do we really need these tag checks here and below, or assertion would be fine? IIRC NULL tag is passed only in `!METASLAB_ASYNC_ALLOC` cases.

Reply:
The `metaslab_alloc_range` call in `zio_write_gang_block` unsets `METASLAB_ASYNC_ALLOC` and passes `NULL` for the tag.

Reply:
Since it unsets, I suppose the check is still not really needed.

Reply:
Sorry, they don't unset `ASYNC_ALLOC`, that's the only flag they leave set.